internal/api/impl/proxy/proxy.go

```diff
@@ -28,67 +29,179 @@ import (
 	"github.com/perses/perses/internal/api/interface/v1/globaldatasource"
 	"github.com/perses/perses/internal/api/interface/v1/globalsecret"
 	"github.com/perses/perses/internal/api/interface/v1/secret"
+	"github.com/perses/perses/internal/api/shared"
 	"github.com/perses/perses/internal/api/shared/crypto"
 	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	"github.com/perses/perses/internal/api/shared/rbac"
 	"github.com/perses/perses/internal/api/shared/route"
 	"github.com/perses/perses/internal/api/shared/utils"
 	v1 "github.com/perses/perses/pkg/model/api/v1"
 	datasourceHTTP "github.com/perses/perses/pkg/model/api/v1/datasource/http"
+	"github.com/perses/perses/pkg/model/api/v1/role"
 	promConfig "github.com/prometheus/common/config"
 	"github.com/sirupsen/logrus"
 )
 
+// unsavedProxyBody is the body of the request when the datasource is not saved yet.
+// It contains the body of the request and the datasource spec, which is used to build the proxy rather than
+// retrieving the datasource from the database.
+type unsavedProxyBody struct {
+	Method string             `json:"method" yaml:"method"`
+	Body   []byte             `json:"body" yaml:"body"`
+	Spec   *v1.DatasourceSpec `json:"spec" yaml:"spec"`
+}
+
+func (u *unsavedProxyBody) setRequestParams(ctx echo.Context) {
+	req := ctx.Request()
+	req.Method = u.Method
+
+	if len(u.Body) > 0 {
+		req.Body = io.NopCloser(strings.NewReader(string(u.Body)))
+	} else {
+		req.Body = nil
+	}
+
+	ctx.SetRequest(req)
+}
+
 type endpoint struct {
 	Dashboard    dashboard.DAO
 	Secret       secret.DAO
 	GlobalSecret globalsecret.DAO
 	DTS          datasource.DAO
 	GlobalDTS    globaldatasource.DAO
 	Crypto       crypto.Crypto
+
+	rbac rbac.RBAC
 }
 
-func New(dashboardDAO dashboard.DAO, secretDAO secret.DAO, globalSecretDAO globalsecret.DAO, dtsDAO datasource.DAO, globalDtsDAO globaldatasource.DAO, crypto crypto.Crypto) route.Endpoint {
+func New(dashboardDAO dashboard.DAO, secretDAO secret.DAO, globalSecretDAO globalsecret.DAO, dtsDAO datasource.DAO, globalDtsDAO globaldatasource.DAO, crypto crypto.Crypto, rbac rbac.RBAC) route.Endpoint {
 	return &endpoint{
 		Dashboard:    dashboardDAO,
 		Secret:       secretDAO,
 		GlobalSecret: globalSecretDAO,
 		DTS:          dtsDAO,
 		GlobalDTS:    globalDtsDAO,
 		Crypto:       crypto,
+
+		rbac: rbac,
+	}
+}
+
+func (e *endpoint) checkPermission(ctx echo.Context, projectName string) error {
```

Comment:
```suggestion
func (e *endpoint) checkPermission(ctx echo.Context, projectName string, scope role.Scope) error {
        claims := crypto.ExtractJWTClaims(ctx)
	if claims == nil {
	  return nil
	}
	
	if role.IsGlobalScope(scope) {
	  [...]
	}
	[....]
}
```

I think you need to pass the scope in the parameter. In the current implementation the scope is hardcoded so even for the `GlobalDatasource` you won't check the accurate permission.

And then in the method `proxyUnsavedGlobalDatasource` you can use it like that 


```go
if err := e.checkPermission(ctx, rbac.GlobalProject, role.GlobalDatasourceScope); err != nil {
		return err
}
```

Reply:
Ah yep - missed that there was a  `KindGlobalDatasource` distinct from `KindDatasource`. Updated  this to take a kind and an action